python_modules/libraries/dagster-airlift/dagster_airlift/core/monitoring_job/event_stream.py

```diff
@@ -0,0 +1,295 @@
+from collections.abc import Iterator
+from itertools import chain
+from typing import Optional
+
+from dagster import AssetMaterialization
+from dagster._core.events import (
+    AssetMaterializationPlannedData,
+    DagsterEvent,
+    JobFailureData,
+    StepMaterializationData,
+)
+from dagster._core.execution.context.op_execution_context import OpExecutionContext
+from dagster._core.remote_representation.origin import RemoteJobOrigin
+from dagster._core.storage.dagster_run import DagsterRun, DagsterRunStatus
+from dagster._core.storage.tags import EXTERNAL_JOB_SOURCE_TAG_KEY
+from dagster._core.utils import make_new_run_id
+from dagster._record import record
+from dagster._time import datetime_from_timestamp
+from dagster_airlift.constants import DAG_ID_TAG_KEY, DAG_RUN_ID_TAG_KEY
+from dagster_airlift.core.airflow_defs_data import AirflowDefinitionsData
+from dagster_airlift.core.airflow_instance import AirflowInstance
+from dagster_airlift.core.monitoring_job.utils import (
+    get_dagster_run_for_airflow_repr,
+    structured_log,
+)
+from dagster_airlift.core.runtime_representations import DagRun, TaskInstance
+
+
+@record
+class DagRunStarted:
+    dag_run: DagRun
+
+    @property
+    def timestamp(self) -> float:
+        return self.dag_run.start_date.timestamp()
+
+    def persist_state(
+        self,
+        context: OpExecutionContext,
+        airflow_data: AirflowDefinitionsData,
+        airflow_instance: AirflowInstance,
+    ) -> None:
+        # If the run is already in Dagster, we don't need to do anything.
+        relevant_job_def = airflow_data.airflow_mapped_jobs_by_dag_handle.get(
+            self.dag_run.dag_handle
+        )
+        # Only if we have a relevant job def do we need to mess with creating runs.
+        if not relevant_job_def:
+            return
+        dagster_run_id = make_new_run_id()
+        context.instance.create_run_for_job(
+            run_id=dagster_run_id,
+            job_def=relevant_job_def,
+            tags={
+                DAG_RUN_ID_TAG_KEY: self.dag_run.run_id,
+                DAG_ID_TAG_KEY: self.dag_run.dag_id,
+                EXTERNAL_JOB_SOURCE_TAG_KEY: "airflow",
+            },
+            status=DagsterRunStatus.NOT_STARTED,
```

Comment:
seems like the run will be started at this point, no? is it just that reporting the PIPELINE_START event later on moves this into the started state?

Reply:
We do this bc the store_event hooks which write out the timestamp to run_stats are required to get the run to actually have a "start time"